Add test coverage for integration with Solidus Starter Frontend

rainerdema opened the pull request:
## Summary

Closes #166

This pull request adds improved test coverage for using the gem with [solidus_starter_frontend](https://github.com/solidusio/solidus_starter_frontend).
The `checkout_spec.rb` has been refactored to include specific use cases with the addition of new scenarios and reorganized test blocks.

The tests cover the following cases:

| Intents flow| Description| Setup Feature Usage | Skip Confirm Step | User | Guest | Payment Result|
|-:|:-|:-:|:-:|:-:|:-:|:-|
| Setup Intent | Credit card with payment capture and wallet reuse | all | no| yes| yes | Successfully created/processed/reused |
| Setup Intent | 3D secure card without capture | all | no| yes | | Required action present |
| Payment Intent | Credit card with payment capture and wallet reuse | all | yes/no | yes| yes | Successfully created/processed/reused |
| Payment Intent | 3D secure card with payment capture | all| yes | yes | | Successfully created/processed |
| Setup Intent | 3D secure card authentication failure | - | -| yes | | Failed authentication |
| Setup Intent | Transaction failure for incorrect data | - | -| yes | | Failed transaction |
| Setup Intent| Incomplete data form submission failure | - | -| yes | | Failed submission |

These tests help ensure the reliability and stability of the gem when used with [solidus_starter_frontend](https://github.com/solidusio/solidus_starter_frontend).

Closes also #187

This PR includes tests that cover all the cards mentioned in the issue ([card that will be declined](https://stripe.com/docs/testing#declined-payments)). 
The issues mentioned in the ticket may be related to previous situations addressed in other PRs:
<details>
    <summary>Refer to the video below to see the current situation in a test environment</summary>
   
  <video src="https://user-images.githubusercontent.com/19948291/224904672-36324941-0a5f-4b2d-9a5a-3049b376cbb7.mov"/>
</details>


## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

## Reviewer Guidelines
To better review the changes made to `checkout_spec.rb`, I suggest reviewing them on the branch rather than the diff.

Here is a link to the relevant file:
[solidus_stripe/spec/system/frontend/solidus_stripe/checkout_spec.rb](https://github.com/solidusio/solidus_stripe/blob/nebulab/rainerd/add-test-coverag-for-the-integration-with-solidus-starter-frontend/spec/system/frontend/solidus_stripe/checkout_spec.rb)
(I can split it into multiple commits if needed)

rainerdema commented:
Please let me know if it would be preferred to split the tests into multiple commits 🙏

elia commented on spec/system/frontend/solidus_stripe/checkout_spec.rb:
```diff
@@ -2,218 +2,302 @@
 require 'solidus_stripe_spec_helper'
 
-RSpec.describe "Checkout with Stripe", :js do
+RSpec.describe 'Checkout', :js do
```
If we're matching the file name maybe we can go one step further and use "SolidusStripe Checkout"?
No big deal, but it's easier to understand where is the problem in case of failures.

elia commented on spec/system/frontend/solidus_stripe/checkout_spec.rb:
```diff
@@ -2,218 +2,302 @@
-      visit_payment_step(user: create(:user))
-      choose_new_stripe_payment
-      fill_stripe_form
+        it 'successfully reuses a previously saved card from the user\'s wallet' do
```
nit: if rubocop is not allowing to use double quotes when we have a single quote inside the string we should change the rule

elia commented on spec/system/frontend/solidus_stripe/checkout_spec.rb:
```diff
@@ -2,218 +2,302 @@
+          # cannot be captured because a required action is needed most of the times.
+          expect_payment_intent_to_be_created_with_required_action
+
+          # TODO: These cards cannot be reused as the required action is not handled.
```
nit: a failing test preceded by a `pending` would be ok here too, not blocking for the approval but if it's super easy to add would be of great help for the next one needing to implement the solution.

rainerdema replied:
Great idea! 👏  I've made the changes:
https://github.com/solidusio/solidus_stripe/blob/759cb54b80ce518a1f97828d29aa4677630d07de/spec/system/frontend/solidus_stripe/checkout_spec.rb#L66

This is the result:
<img width="1550" alt="Screenshot 2023-03-15 at 19 19 28" src="https://user-images.githubusercontent.com/19948291/225406528-dc2bd31d-fe49-4c65-8024-76b64c46a615.png">

Let me know if you were considering adding the `pending` status in the header of the "it" block instead (With the inclusion of the message, it appeared to be more comprehensive to me 🤔).

elia reviewed (approved):
🙌 awesome!

I left a couple of nits for your consideration, they're not blocking tho.

kennyadsl reviewed (approved):
Lovely, thanks @rainerdema!

waiting-for-dev reviewed (approved):
Super-clean job, @rainerdema :clap: :clap: I love how it flows.

I left some non-blocking comments & suggestions. Also, to go for honors, there're some inconsistencies in the method names. Most of them perform both flow + expectations, and some of them are named for the expectation (`expect_declined_cards_to_be_notified`), others like a statement(`successfully_creates_a_setup_intent`), and others like an order (`complete_order`).

waiting-for-dev commented on spec/system/frontend/solidus_stripe/checkout_spec.rb:
```diff
@@ -2,218 +2,302 @@
+  # To learn more about setup_future_usage in different contexts:
+  # - SetupIntents   : https://stripe.com/docs/payments/setup-intents#increasing-success-rate-by-specifying-usage
+  # - PaymentIntents : https://stripe.com/docs/payments/payment-intents#future-usage
+  ['', 'on_session', 'off_session'].each do |setup_future_usage|
```
I guess it's there to be consumed when using the payment intent flow... could we use it only in that case? Otherwise, we have a redundant test for setup intents.

waiting-for-dev commented on spec/support/solidus_stripe/checkout_test_helper.rb:
```diff
@@ -22,12 +22,13 @@ def fill_stripe_form(
     number: 4242_4242_4242_4242, # rubocop:disable Style/NumericLiterals
     expiry_month: 12,
     expiry_year: Time.current.year + 1,
+    date: nil,
```
I got it. That's to simulate an error path.

elia commented on spec/support/solidus_stripe/checkout_test_helper.rb:
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
+      using_wait_time(10) do
+        within_frame(find_stripe_iframe) do
+          field = find(%{input[name="#{name}"]})
+          field.value.length.times { field.send_keys [:backspace] }
```
I can chime in on this one, as it's coming from a previous helper. Basically this is the result of trial and error, and was the most stable way of filling in fields in the stripe iframe, I believe partly because of the JavaScript they run whenever a key is typed, but it's just a hypothesis.

rainerdema replied:
I can confirm that I have tried several solutions, but this is the most stable one I have found.

rainerdema commented on spec/support/solidus_stripe/checkout_test_helper.rb:
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
+  # Submits a 3D Secure payment authorization
+  def authorize_3d_secure_payment(authenticate: true)
+    find_frame('body > div > iframe') do
+      sleep 1
```
I'll make a few tries, but probably yes (and it's already in the `find_frame` helper method)

rainerdema commented on spec/system/frontend/solidus_stripe/checkout_spec.rb:
```diff
@@ -2,218 +2,302 @@
-  def submit_payment
-    click_button("Save and Continue")
-    expect(page).to have_content("Agree to Terms of Service")
+  private
```
For no particular reason 😄 Upon re-reading the specs, I noticed that these methods were the primary ones most used for constructing the various test blocks. As a result, I decided to keep them easily accessible within the `checkout_spec.rb` file. Anyway, I'll move them all to the dedicated helper for consistency 👍

rainerdema commented on spec/system/frontend/solidus_stripe/checkout_spec.rb:
```diff
@@ -2,218 +2,302 @@
+      fill_stripe_form(number: '4242424242424241')
+      submit_payment
+      using_wait_time(10) do
+        within_frame(find_stripe_iframe) do
```
Great suggestion! I've moved this part to a dedicated method:
https://github.com/solidusio/solidus_stripe/blob/759cb54b80ce518a1f97828d29aa4677630d07de/spec/support/solidus_stripe/checkout_test_helper.rb#L242-L259

rainerdema commented:
@elia @kennyadsl @waiting-for-dev 👋 Thanks everyone for the reviews!
I have looked over your suggestions and made the necessary revisions.

Here are a few considerations:
- Considering the number of methods in the [checkout_test_helper.rb](https://github.com/solidusio/solidus_stripe/blob/90c4482934057b90219afeed84290ff13305fdf9/spec/support/solidus_stripe/checkout_test_helper.rb), I have reorganized and grouped them as efficiently as possible.
- I'm waiting for the merge of #217 to integrate the retrieve methods into the helper before merging (https://github.com/solidusio/solidus_stripe/pull/215#discussion_r1136709134)
- I have removed the static `off_session` setting from the setup intent creation to ensure that the tests correctly cover all configuration flows: https://github.com/solidusio/solidus_stripe/pull/215/commits/90c4482934057b90219afeed84290ff13305fdf9
Let me know if we're ok with that 👍

rainerdema commented:
> there're some inconsistencies in the method names. Most of them perform both flow + expectations, and some of them are named for the expectation (expect_declined_cards_to_be_notified), others like a statement(successfully_creates_a_setup_intent), and others like an order (complete_order).

@waiting-for-dev Yes, I agree with you that In `expect_declined_cards_to_be_notified`, `successfully_creates_a_setup_intent` and `complete_order` for example, there are expectation instructions that appear to be mixed in with flow instructions.
I'm open to suggestions in case the method nomenclature can be better improved 👍

However, most of the time, these checks are necessary to ensure that the subsequent commands are executed at the appropriate time.
Let me know if this makes sense and if it addresses your concerns.

waiting-for-dev commented:
> However, most of the time, these checks are necessary to ensure that the subsequent commands are executed at the appropriate time.
> Let me know if this makes sense and if it addresses your concerns.

Yeah, I'm not suggesting removing the checks. Just using the same naming convention. For instance, for the three examples:

- `declined_cards_are_notified`
- `successfully_creates_a_setup_intent`
- `completes_order`

I'm picking the statement format because it's the more neutral between `expect_`, which looks like it only contains RSpec expectation`, and the imperative, which looks like it only contains flow instructions.

But leaving it to your chose!
